### PR TITLE
ddl: Speed up adding index phase

### PR DESCRIPTION
**Pull request by zimulala**
Handle issue #2257
Remove the batch get operation when adding index.

Using 3 TiKV, 1 PD, 1 TiDB , the servers of TiKV and TiDB are not on a computer, two TiKV on a computer.
The lease of TiDB is 1s.
The number of records in this table is 3531200(3.4 M), and the table's structure is as follows:
+------------------+------------------+------+------+---------+-------+
| Field | Type | Null | Key | Default | Extra |
+------------------+------------------+------+------+---------+-------+
| a | varchar(64) | YES | | | |
| b | bigint(20) | NO | | 0 | |
| c | datetime | YES | | NULL | |
| d | int(11) UNSIGNED | NO | | 0 | |
| e | int(11) UNSIGNED | NO | | 0 | |
| f | int(11) UNSIGNED | NO | | 0 | |
| g | varchar(28) | NO | | | |
| h | varchar(28) | NO | | | |
+------------------+------------------+------+------+---------+-------+

tidb-before> alter table battle_begin add index b (event_id);
Query OK, 0 rows affected (8 min 36.40 sec)
tidb-before>  alter table battle_begin add index g (sum_id);
Query OK, 0 rows affected (7 min 10.01 sec)

tidb-current> alter table battle_begin add index b (event_id);
Query OK, 0 rows affected (3 min 35.75 sec)
tidb-current> alter table battle_begin add index g (sum_id);
Query OK, 0 rows affected (3 min 46.25 sec)

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -324,43 +326,99 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
 	return errors.Trace(err)
 }
 
-func (d *ddl) fetchRowColVals(txn kv.Transaction, t table.Table, batchOpInfo *indexBatchOpInfo, seekHandle int64) error {
-	cols := t.Cols()
-	idxInfo := batchOpInfo.tblIndex.Meta()
+func isFinish(limit, input int64) bool {
```
Comment (coocood) [inline ddl/index.go]: isFinished

File: ddl/index.go [hunk 2]
```diff
@@ -375,120 +433,137 @@ func (d *ddl) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgInfo
...
 	seekHandle := reorgInfo.Handle
+	wg := sync.WaitGroup{}
 	for {
 		startTime := time.Now()
-		err := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
-			err1 := d.isReorgRunnable(txn, ddlJobFlag)
-			if err1 != nil {
-				return errors.Trace(err1)
+		for i := 0; i < batches; i++ {
+			go d.backfillIndex(t, batchOpInfo, seekHandle, &wg)
```
Comment (coocood) [inline ddl/index.go]: We should Add waitgroup outside the goroutine

File: ddl/index.go [hunk 3]
```diff
@@ -375,120 +433,137 @@ func (d *ddl) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgInfo
...
 		colMap:     colMap,
-		handle:     reorgInfo.Handle,
-		idxRecords: make([]*indexRecord, 0, batchCnt),
+		nextCh:     make(chan int64, 1),
+		batchRetCh: make(chan *batchRet, batches),
 	}
 
+	addedCount := job.GetRowCount()
 	seekHandle := reorgInfo.Handle
+	wg := sync.WaitGroup{}
```
Comment (coocood) [inline ddl/index.go]: Put this in for loop is better.

File: ddl/index.go [hunk 4]
```diff
@@ -324,43 +326,99 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+	err        error
+}
+
+type batchRetSlice []*batchRet
+
+func (b batchRetSlice) Len() int           { return len(b) }
+func (b batchRetSlice) Less(i, j int) bool { return b[i].doneHandle < b[j].doneHandle }
+func (b batchRetSlice) Swap(i, j int)      { b[i], b[j] = b[j], b[i] }
+
+type indexRecord struct {
```
Comment (shenli) [inline ddl/index.go]: Please add  comments for the following structs.

File: ddl/index.go [hunk 5]
```diff
@@ -324,43 +326,100 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+}
+
+const (
+	defaultBatchCnt      = 1024
+	defaultSmallBatchCnt = 128
+	defaultSmallBatches  = 16
+)
+
+// batchRet is the result of the batch.
+type batchRet struct {
```
Comment (coocood) [inline ddl/index.go]: rename to `batchResult` is better.

File: ddl/index.go [hunk 6]
```diff
@@ -324,43 +326,100 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
 	return errors.Trace(err)
 }
 
-func (d *ddl) fetchRowColVals(txn kv.Transaction, t table.Table, batchOpInfo *indexBatchOpInfo, seekHandle int64) error {
-	cols := t.Cols()
-	idxInfo := batchOpInfo.tblIndex.Meta()
+func isFinished(limit, input int64) bool {
```
Comment (coocood) [inline ddl/index.go]: The function name is too generic.
make it as a method of `handleInfo` is better.

File: ddl/index.go [hunk 7]
```diff
@@ -324,43 +326,100 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+	if limit == 0 || input < limit {
```
Comment (coocood) [inline ddl/index.go]: A 0 handle can be valid.

File: ddl/index.go [hunk 8]
```diff
@@ -324,43 +326,100 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
-	count := len(batchOpInfo.idxRecords)
-	batchOpInfo.addedCount += int64(count)
-	batchOpInfo.handle = batchOpInfo.idxRecords[count-1].handle
-	return nil
+	ret.count = len(idxRecords)
+	if ret.count > 0 {
+		ret.doneHandle = idxRecords[ret.count-1].handle
+	}
+	// Be sure to do this operation only once.
+	handleInfo.once.Do(func() {
```
Comment (coocood) [inline ddl/index.go]: We don't have to use `once`, add a bool member in handleInfo is clearer.

File: ddl/index.go [hunk 9]
```diff
@@ -375,120 +434,137 @@ func (d *ddl) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgInfo
...
+			wg.Add(1)
+			go d.backfillIndex(t, batchOpInfo, seekHandle, &wg)
+			handle := <-batchOpInfo.nextCh
+			// There is no data to seek.
+			if handle == 0 {
+				break
 			}
-			batchOpInfo.idxRecords = batchOpInfo.idxRecords[:0]
-			err1 = d.backfillIndexInTxn(t, txn, batchOpInfo, seekHandle)
+			seekHandle = handle + 1
```
Comment (coocood) [inline ddl/index.go]: rename `seekHandle` to `batchStartHandle` is better.

File: ddl/index.go [hunk 10]
```diff
@@ -375,120 +434,137 @@ func (d *ddl) addTableIndex(t table.Table, indexInfo *model.IndexInfo, reorgInfo
...
 	for {
 		startTime := time.Now()
-		err := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
-			err1 := d.isReorgRunnable(txn, ddlJobFlag)
-			if err1 != nil {
-				return errors.Trace(err1)
+		for i := 0; i < batches; i++ {
+			wg.Add(1)
+			go d.backfillIndex(t, batchOpInfo, seekHandle, &wg)
+			handle := <-batchOpInfo.nextCh
```
Comment (coocood) [inline ddl/index.go]: name it `doneHandle`, keep it consistent with the sender's

File: ddl/index.go [hunk 11]
```diff
@@ -324,43 +326,100 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
-const defaultSmallBatchCnt = 128
+type indexBatchOpInfo struct {
+	tblIndex   table.Index
+	colMap     map[int64]*types.FieldType
+	batchRetCh chan *batchRet
+	nextCh     chan int64 // It notifies to start the next batch.
+}
 
 // How to add index in reorganization state?
 //  1. Generate a snapshot with special version.
```
Comment (coocood) [inline ddl/index.go]: Update the comment.

**Comment by ngaut** [comment]
PTAL

**Comment by zimulala** [comment]
PTAL @coocood

File: ddl/index.go [hunk 12]
```diff
@@ -418,11 +418,23 @@ type indexBatchOpInfo struct {
...
+//  1. Open a goroutine. Traverse the snapshot to obtain the handle range, while access to the corresponding row key and
+// raw index value. Then notify to start the next batch.
+//  2. Decoding this batch of raw index value, gets the corresponding index value.
+//  3. Deal with this index records one by one. If the index record is existed, skip to the next row.
+// If the index doesn't exist, create the index ande then continue to handle the next row.
+//  4. When the handle of range is completed, returns the corresponding batch result.
+// The above operations are completed in a transaction.
+// When concurrent tasks are processed, the batch result returned by each batch is sorted by handle. Then traverse the
+// batch results, gets the total number of row in the concurrent task and update the processed handle value. If
+// you encounter an error message, exit traversal.
```
Comment (coocood) [inline ddl/index.go]: s/you/we

File: ddl/index.go [hunk 13]
```diff
@@ -324,171 +326,261 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+	nextCh     chan int64                 // It notifies to start the next batch.
+}
 
 // How to add index in reorganization state?
-//  1. Generate a snapshot with special version.
-//  2. Traverse the snapshot, get every row in the table.
-//  3. For one row, if the row has been already deleted, skip to next row.
-//  4. If not deleted, check whether index has existed, if existed, skip to next row.
-//  5. If index doesn't exist, create the index and then continue to handle next row.
+// Concurrently process defaultSmallBatches tasks. Each task deals with a handle interval of the index record.
```
Comment (coocood) [inline ddl/index.go]: we can use `handle range` uniformly.

File: ddl/index.go [hunk 14]
```diff
@@ -324,171 +326,261 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+// The handle interval is defaultSmallBatchCnt.
```
Comment (coocood) [inline ddl/index.go]: handle range size

File: ddl/index.go [hunk 15]
```diff
@@ -324,171 +326,261 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+// Although the length of each handle interval is controllable, but the range of the handle value can't be expected,
```
Comment (coocood) [inline ddl/index.go]: Because each handle range depends on the previous one, it's necessary to obtain the handle range sequentially.

**Comment by zimulala** [comment]
PTAL @ngaut @coocood

File: ddl/index.go [hunk 16]
```diff
@@ -324,171 +326,260 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+	}
+	return idxRecords, ret
 }
 
-const defaultBatchCnt = 1024
-const defaultSmallBatchCnt = 128
+const (
+	defaultBatchCnt      = 1024
+	defaultSmallBatchCnt = 128
+	defaultSmallBatches  = 16
```
Comment (coocood) [inline ddl/index.go]: defaultTasks

File: ddl/index.go [hunk 17]
```diff
@@ -324,171 +326,260 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+)
+
+// batchResult is the result of the batch.
+type batchResult struct {
```
Comment (coocood) [inline ddl/index.go]: taskResult

File: ddl/index.go [hunk 18]
```diff
@@ -324,171 +326,260 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+	nextCh     chan int64                 // It notifies to start the next batch.
+}
 
 // How to add index in reorganization state?
-//  1. Generate a snapshot with special version.
-//  2. Traverse the snapshot, get every row in the table.
-//  3. For one row, if the row has been already deleted, skip to next row.
-//  4. If not deleted, check whether index has existed, if existed, skip to next row.
-//  5. If index doesn't exist, create the index and then continue to handle next row.
+// Concurrently process defaultSmallBatches tasks. Each task deals with a handle range of the index record.
```
Comment (queenypingcap) [inline ddl/index.go]: defaultSmallBatches tasks -> the defaultSmallBatches tasks.

File: ddl/index.go [hunk 19]
```diff
@@ -324,171 +326,260 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+// The handle range size is defaultSmallBatchCnt.
+// Because each handle range depends on the previous one, it's necessary to obtain the handle range serially.
+// Real concurrent processing needs to perform after the handle range has been acquired.
+// The operation flow of the each batch of data is as follows:
+//  1. Open a goroutine. Traverse the snapshot to obtain the handle range, while access to the corresponding row key and
```
Comment (queenypingcap) [inline ddl/index.go]: access to ->accessing

File: ddl/index.go [hunk 20]
```diff
@@ -324,171 +326,260 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+// raw index value. Then notify to start the next batch.
+//  2. Decoding this batch of raw index value gets the corresponding index value.
```
Comment (queenypingcap) [inline ddl/index.go]: Decoding this batch of raw index value gets the corresponding index value. ->Decode this batch of raw index value to get the corresponding index value.

File: ddl/index.go [hunk 21]
```diff
@@ -324,171 +326,260 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+//  3. Deal with this index records one by one. If the index record exists, skip to the next row.
```
Comment (queenypingcap) [inline ddl/index.go]: this index records -> these index records

File: ddl/index.go [hunk 22]
```diff
@@ -324,171 +326,260 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+// If the index doesn't exist, create the index ande then continue to handle the next row.
+//  4. When the handle of a range is completed, returns the corresponding batch result.
```
Comment (queenypingcap) [inline ddl/index.go]: returns ->return

File: ddl/index.go [hunk 23]
```diff
@@ -324,171 +326,260 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+// The above operations are completed in a transaction.
+// When concurrent tasks are processed, the batch result returned by each batch is sorted by the handle. Then traverse the
+// batch results, gets the total number of row in the concurrent task and update the processed handle value. If
```
Comment (queenypingcap) [inline ddl/index.go]: gets -> get

File: ddl/index.go [hunk 24]
```diff
@@ -324,171 +326,260 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+// we encounter an error message, exit traversal.
```
Comment (queenypingcap) [inline ddl/index.go]: -> an error message is displayed, exit the traversal.

File: ddl/index.go [hunk 25]
```diff
@@ -324,184 +326,275 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
 		colMap[col.ID] = &col.FieldType
 	}
-	batchCnt := defaultSmallBatchCnt
-	batchOpInfo := &indexBatchOpInfo{
-		tblIndex:   tables.NewIndex(t.Meta(), indexInfo),
-		addedCount: job.GetRowCount(),
-		colMap:     colMap,
-		handle:     reorgInfo.Handle,
-		idxRecords: make([]*indexRecord, 0, batchCnt),
+	tasks := defaultTaskCnt
```
Comment (coocood) [inline ddl/index.go]: taskCnt

File: ddl/index.go [hunk 26]
```diff
@@ -324,184 +326,275 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+		if retCnt < tasks {
 			return nil
 		}
-		seekHandle = batchOpInfo.handle + 1
 	}
 }
 
-// recordIterFunc is used for low-level record iteration.
-type recordIterFunc func(h int64, rowKey kv.Key, rawRecord []byte) (more bool, err error)
+// handleInfo records start ande end handle that is used in a task.
```
Comment (coocood) [inline ddl/index.go]: and

**Comment by coocood** [comment]
LGTM

File: ddl/index.go [hunk 27]
```diff
@@ -324,184 +326,275 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
-	defer it.Close()
+	sort.Sort(taskRetSlice(taskRets))
 
-	for it.Valid() {
-		if !it.Key().HasPrefix(t.RecordPrefix()) {
+	taskAddedCount, currHandle := int64(0), int64(0)
+	var err error
+	for _, ret := range taskRets {
+		if ret.err != nil {
+			err = ret.err
```
Comment (hanfei1991) [inline ddl/index.go]: return directly here? then u don't need to use an err var.

File: ddl/index.go [hunk 28]
```diff
@@ -324,184 +326,275 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+		if taskAddedCount != 0 {
+			err1 := kv.RunInNewTxn(d.store, true, func(txn kv.Transaction) error {
+				return errors.Trace(reorgInfo.UpdateHandle(txn, doneHandle+1))
+			})
 			if err1 != nil {
-				return errors.Trace(err1)
+				if err == nil {
+					err = err1
+				} else {
+					log.Warnf("[ddl] add index failed when update handle %d, err %v", doneHandle, err)
```
Comment (hanfei1991) [inline ddl/index.go]: use err %s and err.Error() ?

Comment (zimulala) [inline ddl/index.go]: This isn't necessary.  Using `err` is clearer.

File: ddl/index.go (hunk 27 above)
Comment (zimulala) [inline ddl/index.go]: I think that's OK.

File: ddl/index.go [hunk 29]
```diff
@@ -324,184 +326,275 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
-			return errors.Trace(err)
+			taskRet.err = errors.Trace(err)
+			return taskRet
 		}
 
 		// Create the index.
-		handle, err := batchOpInfo.tblIndex.Create(txn, idxRecord.vals, idxRecord.handle)
+		handle, err := taskOpInfo.tblIndex.Create(txn, idxRecord.vals, idxRecord.handle)
 		if err != nil {
 			if terror.ErrorEqual(err, kv.ErrKeyExists) && idxRecord.handle == handle {
```
Comment (hanfei1991) [inline ddl/index.go]: When it returns err key exists but the handle is not the same?

File: ddl/index.go [hunk 30]
```diff
@@ -324,184 +326,275 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
-	count := len(batchOpInfo.idxRecords)
-	batchOpInfo.addedCount += int64(count)
-	batchOpInfo.handle = batchOpInfo.idxRecords[count-1].handle
-	return nil
+	ret.count = len(idxRecords)
+	if ret.count > 0 {
+		ret.doneHandle = idxRecords[ret.count-1].handle
+	}
+	// Be sure to do this operation only once.
+	if !handleInfo.isSent {
```
Comment (hanfei1991) [inline ddl/index.go]: When the handleInfo's `isSent` is true at this time?

File: ddl/index.go [hunk 31]
```diff
@@ -324,184 +326,275 @@ func (d *ddl) onDropIndex(t *meta.Meta, job *model.Job) error {
...
+		// Notice to start the next task operation.
+		taskOpInfo.nextCh <- ret.doneHandle
+		// Record the last handle.
+		// Ensure that the handle scope of the task doesn't change,
+		// even if the transaction retries it can't effect the other tasks.
+		handleInfo.endHandle = ret.doneHandle
+		handleInfo.isSent = true
+	}
+	if ret.count == 0 {
```
Comment (hanfei1991) [inline ddl/index.go]: Does `ret.count` always equal to `len(idxRecords)`? If so, you needn't record it.

File: ddl/index.go (hunk 29 above)
Comment (zimulala) [inline ddl/index.go]: It's a unique key, the key value is updated when we create the index.

File: ddl/index.go (hunk 28 above)
Comment (hanfei1991) [inline ddl/index.go]: Then should you log err1 in this place ?

File: ddl/index.go (hunk 31 above)
Comment (zimulala) [inline ddl/index.go]: Yes. But we need it to update statistics.

**Pull request merged**